app/controllers/revisions_controller.rb

```diff
@@ -40,6 +40,44 @@ def show
             )
   end
 
+  def trash
+    # Get all versions whose event is destroy.
+    @revisions = PaperTrail::Version.where(event: 'destroy').order(created_at: :desc)
+  end
+
+  def recover
+    revision = PaperTrail::Version.find params[:id]
+    object = revision.reify
+    # If object's node was destroyed, assign it no a new node.
+    if !Node.exists?(object.node_id)
+      recovered_node = Node.create(label: 'Recovered', type_id: Node::Types::DEFAULT)
+      object.node_id = recovered_node.id
+    end
+
+    # If object is evidence and its issue doesn't exist any more, recover it.
+    if revision.item_type == 'Evidence' and !Note.exists?(object.issue_id)
+      issue_revision = PaperTrail::Version.where(event: 'destroy', item_type: 'Note', item_id: object.issue_id).limit(1).first
+      # A destroy revision should always be present, but just in case.
+      if issue_revision
+        issue_object = issue_revision.reify
+        issue_object.node_id = Node.issue_library.id
+        issue_object.save
+        object.issue_id = issue_object.id
+        issue_revision.destroy
+      end
+    end
+
+    if object.save
+      # Destroy revision so item is not listed in trash any more.
+      revision.destroy
```

Comment:
Here's the best I can come up with:

    ids = [Evidence, Note].flat_map do |model|
      table_name = model.table_name
      PaperTrail::Version.where(event: 'destroy', item_type: model.to_s).
        joins("LEFT JOIN #{table_name} ON item_id=#{table_name}.id").
        where("#{table_name}.id IS NULL"). # avoid showing deleted objects
        select("distinct item_id, versions.id").
        pluck("versions.id")
    end
    PaperTrail::Version.where(id: ids)

But this isn't ideal because it makes multiple SQL queries. Do you think this is okay?

For posterity's sake, here's something I tried that didn't work (although I can't figure out why):

    PaperTrail::Version.where(event: 'destroy').
      joins("LEFT JOIN #{Note.table_name} ON item_id=#{Note.table_name}.id").
      joins("LEFT JOIN #{Evidence.table_name} ON item_id=#{Evidence.table_name}.id").
      where("#{Note.table_name}.id IS NULL").
      where("#{Evidence.table_name}.id IS NULL").
      select("distinct item_id, versions.*")